### PR TITLE
ci: automate pinning charms in regression tests

### PR DESCRIPTION
**Pull request by dimaqq**
Switch to using pinned versions of the charm (latest main) for our charm tests in the GitHub Actions.

- [x] charmcraft-pack.yaml
- [x] db-charm-tests.yaml
- [x] ~framework-tests.yaml~ no change needed
- [x] hello-charm-tests.yaml
- [x] observability-charm-tests.yaml
- [x] ~publish.yml~ aggregates other workflows
- [x] ~test-publish.yml~ aggregates other workflows
- [x] automation to update pins
   - [x] detect changes
   - [x] create PR to update pins
   - [x] setup, token 

Also I've re-enabled mysql-k8s charm test, as that apparently got fixed upstream.

Here's the configuration for the personal access token I'm using to develop the external charm "dependabot":
- read/write access to the repo
- `workflow` scope to push branches with changes to workflows

Screenshot 2024-05-23 at 17 07 50

### CODE REVIEW

**Comment by benhoyt** [comment]
Yeah, something like that -- or should we pin the latest commit hashes on their main/master? @tonyandrewmeyer @IronCore864

**Comment by dimaqq** [comment]
They seem to tag latest on `main` with these tags, here's `git log main` for the prometheus charm:

```Git Revision List
commit 091c41a49182826539eb4771f0463cd805554453 (tag: rev186, origin/main, origin/HEAD, main)
Author: Noctua <webops+observability-noctua-bot@canonical.com>
Date:   Sat May 4 02:03:24 2024 +0200

    chore: update charm libraries (#603)

commit 69b500c3d331bc1c35ac1c0e650c775e922410c4 (tag: rev185)
Author: Noctua <webops+observability-noctua-bot@canonical.com>
Date:   Fri May 3 22:01:14 2024 +0200

    chore: update charm libraries (#602)

commit d35f2ea680a6164d31ff3935a8f4bc9bda9ad7d3 (tag: rev184)
Author: Noctua <webops+observability-noctua-bot@canonical.com>
Date:   Fri May 3 02:03:16 2024 +0200

    chore: update charm libraries (#601)

commit fcfc555317058f0dae1f5fc73861c79c5415e9f9 (tag: rev183)
Author: Noctua <webops+observability-noctua-bot@canonical.com>
Date:   Thu May 2 22:01:10 2024 +0200

    chore: update charm libraries (#598)

commit 9077888faf9d8a8fc337d34c042ec1aa5e09a70c (tag: rev182, origin/external-url-in-datasource)
Author: Leon <82407168+sed-i@users.noreply.github.com>
Date:   Tue Apr 30 09:09:49 2024 -0400

    Fix sanitation (#599)

```

**Comment by benhoyt** [comment]
Okay, per team discussion, for observability charms we'll use the always-updated "revNNN" tags. For other teams' charms that don't always update the tag we'll use the latest commit hash.

File: .github/workflows/db-charm-tests.yaml [hunk 1]
```diff
@@ -8,19 +8,22 @@ jobs:
...
-        charm-repo:
-          - "canonical/postgresql-operator"
-          - "canonical/postgresql-k8s-operator"
-          - "canonical/mysql-operator"
-# TODO: uncomment once secrets issues are fixed in this charm:
-#       https://github.com/canonical/mysql-k8s-operator/pull/371
-#          - "canonical/mysql-k8s-operator"
+        include:
+          - charm-repo: canonical/postgresql-operator
+            tag: rev405
```
Comment (tonyandrewmeyer) [inline .github/workflows/db-charm-tests.yaml:13]: It seems like there are some commits (e.g. 82e6cb91cae5123e86ffc24d24a1e1a2fc1f8043) that don't get tags, maybe ones done by bots? Wouldn't it be better to just go with whatever the latest hash is?

File: .github/workflows/charmcraft-pack.yaml [hunk 2]
```diff
@@ -11,6 +11,7 @@ jobs:
       uses: actions/checkout@v3
       with:
         repository: jnsgruk/hello-kubecon
+        ref: dbd133466dde59ee64f20a732a8f3d2e560ec3b8  # master, July 3rd, 2023
```
Comment (tonyandrewmeyer) [inline .github/workflows/charmcraft-pack.yaml:14]: I don't think we need to have the branch name in the comment, since we'll always be using whatever the default branch is - it also avoids adding "master" into our repo. I do like the date stamp, though - but if we do that then we should do it for all of them for consistency.

File: .gitignore [hunk 3]
```diff
@@ -9,6 +9,9 @@ venv
 .vscode
 .coverage
 /.tox
+.env
+.envrc
+.secrets
```
Comment (tonyandrewmeyer) [inline .gitignore:14]: I have no objection to adding these to .gitignore, and generally no objection to small drive-bys. However, I think for .gitignore it's cleaner if changes are in dedicated (or related) PRs.

**Review by tonyandrewmeyer — CHANGES_REQUESTED** [review]
Thanks! Looks good, just a couple of small comments.

File: .github/workflows/hello-charm-tests.yaml [hunk 4]
```diff
@@ -22,6 +24,7 @@ jobs:
         uses: actions/checkout@v3
         with:
           repository: ${{ matrix.charm-repo }}
+          ref: ${{ matrix.revision }}
```
Comment (benhoyt) [inline .github/workflows/hello-charm-tests.yaml:27]: This is "tag" some places and "revision" others. But we may sometimes set "tag" to a revision. I think we should just use "ref" everywhere.

File: .gitignore (hunk 3 above)
Comment (benhoyt) [reply .gitignore:14]: Either works for me. If you decide to keep them here, can you please add a PR comment as to why these are needed?

**Review by benhoyt — CHANGES_REQUESTED** [review]

**Comment by dimaqq** [comment]
After Madrid conversations, I'm going to change pins to git commit hashes, each with a comment (tag?, date).

File: .github/workflows/hello-charm-tests.yaml (hunk 4 above)
Comment (dimaqq) [reply .github/workflows/hello-charm-tests.yaml]: I've changed the pins back to commits

File: .gitignore (hunk 3 above)
Comment (dimaqq) [reply .gitignore:16]: Done

File: .github/workflows/charmcraft-pack.yaml (hunk 2 above)
Comment (dimaqq) [reply .github/workflows/charmcraft-pack.yaml:14]: Reworked.

File: .github/workflows/db-charm-tests.yaml (hunk 1 above)
Comment (dimaqq) [reply .github/workflows/db-charm-tests.yaml]: I've changed pins back to commits

**Comment by dimaqq** [comment]
Example PRs that automation generates:
- test dummy, closed by automation: https://github.com/dimaqq/operator/pull/17
- opened by automation, manually merged: https://github.com/dimaqq/operator/pull/19

File: .github/workflows/update-charm-tests.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          PR=$(gh pr list --state open --head auto-update-external-charm-pins --json number -q '.[0].number')
+          CHANGES=$(git --no-pager diff --stat main HEAD)
+          echo "Existing PR? $PR"
+          echo "Changes? $CHANGES"
+          if [[ -n "$PR" && -z "$CHANGES" ]]; then
+            echo "Closing #$PR as stale"
+            gh pr close -c stale "$PR";
+          elif [[ -z "$PR" && -n "$CHANGES" ]]; then
+            echo "Opening new PR"
+            gh pr create --base main --head auto-update-external-charm-pins --title "Update Charm Pins" --body "This is an automated PR to update pins of the external repositories that the operator framework is tested against";
```
Comment (dimaqq) [inline .github/workflows/update-charm-tests.yaml:50]: I've now noticed that the ops repo follows conventional commit naming, in lower case so maybe I should change the PR title and the commit message to `chore: update charm pins` ?

Comment (benhoyt) [reply .github/workflows/update-charm-tests.yaml:50]: Yep, agreed, thanks.

**Review by benhoyt — APPROVED** [review]
Looks reasonable to me -- thanks!

File: .github/actions/update-charm-pins/main.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# Copyright 2024 Canonical Ltd.
+
+"""Updates pinned versions of charms in tests."""
+
+import logging
+import os
+import sys
+
+from httpx import Client
+from ruamel.yaml import YAML
```
Comment (tonyandrewmeyer) [inline .github/actions/update-charm-pins/main.py:10]: Are you using ruamel.yaml rather than pyyaml (which ops itself uses) from personal preference, or does it have functionality that's used here that pyyaml is missing? This is only a CI utility script, but in general keeping the number of packages down is good.

File: .github/actions/update-charm-pins/action.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    required: true
+
+runs:
+  using: composite
+  steps:
+    - uses: actions/setup-python@v5
+      with:
+        python-version: "3.12"
+
+    - run: python -m pip install ruamel.yaml==0.18.6 httpx==0.27.0
```
Comment (tonyandrewmeyer) [inline .github/actions/update-charm-pins/action.yaml:24]: Do you know if dependabot is smart enough to find these? If not, then it's probably better to have a requirements file and install from that.

**Review by tonyandrewmeyer — CHANGES_REQUESTED** [review]
Looks good - a couple of questions and couple of nits.

I do think it would be worthwhile documenting how to run this manually in HACKING.md (unless that's going to be done in #1213). Just a short section that explains that this workflow/action exists, and how to run it (potentially triggering the action with GitHub's UI/CLI, or with `act`, but also just the basic `GITHUB_TOKEN=github_xxxx python3 main.py ../../workflows/hello-charm-tests.yaml` approach (which is the extent of local testing that I did for the review, fwiw).

File: .github/workflows/update-charm-tests.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
+---
+name: Update Charm Pins
+
+on:
+  # NOTE: to avoid infinite loop, exclude the branch created by this workflow if triggering on push or pull_request
+  workflow_dispatch:
+  schedule:
+    - cron: '0 18 * * SUN' # Sunday 6pm UTC, before international day starts
```
[the remark this answers is not in the document]
Comment (dimaqq) [reply .github/workflows/update-charm-tests.yaml:8]: I was thinking that it would be to the team.

In the current state:
- weekly runs are automatic
- manual trigger works, but only from the main branch.*
- the Python script can be ran manually to update current local branch

*) Since the generated PR targets the main branch, I figured that limitation was reasonable.

File: .github/actions/update-charm-pins/main.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    # Assume the workflow has a single job or the first job is parameterized with charm repos
+    job_name = next(iter(doc["jobs"]))
+
+    for idx, item in enumerate(doc["jobs"][job_name]["strategy"]["matrix"]["include"]):
+        charm_repo = item["charm-repo"]
+
+        resp = github.get(f"{charm_repo}/commits", params={"per_page": 1})
+        resp.raise_for_status()
+        commit = resp.json()[0]["sha"]
+        timestamp = resp.json()[0]["commit"]["committer"]["date"]
```
[the remark this answers is not in the document]
Comment (dimaqq) [reply .github/actions/update-charm-pins/main.py:38]: Ah! I never realised that `.raise_for_status` supports chaining... Or maybe I was misled by a method with the same name in some other library. Thank you!

File: .github/actions/update-charm-pins/main.py (hunk 6 above)
Comment (dimaqq) [reply .github/actions/update-charm-pins/main.py:10]: 1. for (limited) round-trip safety
2. to insert comments

File: .github/actions/update-charm-pins/main.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dimaqq) [reply .github/actions/update-charm-pins/main.py]: done

File: .github/actions/update-charm-pins/main.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    base_url="https://api.github.com/repos",
+    headers={
+        "Authorization": f'token {os.getenv("GITHUB_TOKEN")}',
+        "Accept": "application/vnd.github.v3+json",
+    },
+)
+
+
+def update_charm_pins(workflow):
+    """Update pinned versions of charms in the given github actions workflow."""
```
[the remark this answers is not in the document]
Comment (dimaqq) [reply .github/actions/update-charm-pins/main.py]: done

File: .github/workflows/update-charm-tests.yaml (hunk 5 above)
Comment (dimaqq) [reply .github/workflows/update-charm-tests.yaml]: done

**Comment by dimaqq** [comment]
I've added a block to hacking and wrote a readme for the action.

**Comment by benhoyt** [comment]
@dimaqq Is https://github.com/canonical/operator/pull/1213 still relevant given the weekly automation in this PR?

File: .github/actions/update-charm-pins/main.py (hunk 6 above)
Comment (tonyandrewmeyer) [reply .github/actions/update-charm-pins/main.py:10]: Seems reasonable. Perhaps at some point we should consider using it for ops as well? It's a little icky using two packages that do (more-or-less) the same thing, and if this is better, it might be worth it. MPL2 might be an issue in ops itself, though,

File: HACKING.md [hunk 11]
```diff
@@ -155,6 +155,11 @@ your charm to a controller using that version of Juju. For example, with microk8
 3. Run `GOBIN=/path/to/your/juju/_build/linux_amd64/bin:$GOBIN /path/to/your/juju bootstrap`
 4. Add a model and deploy your charm as normal
 
+### Regression testing against existing charms
+
+We rely on automation to [update charm pins](.github/actions/update-charm-pins/) of
+a bunch of charms that use the operator framework. The script can be ran locally too.
```
Comment (tonyandrewmeyer) [inline HACKING.md:161]:
```suggestion
a bunch of charms that use the operator framework. The script can be run locally too.
```

**Review by tonyandrewmeyer — APPROVED** [review]
Nice!

File: .github/actions/update-charm-pins/main.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tonyandrewmeyer) [reply .github/actions/update-charm-pins/main.py]: > Ah! I never realised that `.raise_for_status` supports chaining... Or maybe I was misled by a method with the same name in some other library. Thank you!

Probably requests - this is a nice improvement that httpx made over requests, I believe - I'm pretty sure that `raise_for_status` in requests doesn't have this behaviour.

File: .github/actions/update-charm-pins/main.py (hunk 6 above)
Comment (dimaqq) [reply .github/actions/update-charm-pins/main.py:10]: True, though it seems that 2-packages approach is kindof the way forward.

See e.g. https://peps.python.org/pep-0680/#including-an-api-for-writing-toml

**Comment by dimaqq** [comment]
I'll update #1213
